Run deployer pods with service account credentials

liggitt opened the pull request:
Fixes #2506

- [x] Removes the cluster-wide deployer user (cert generation, config, etc)
- [x] Adds "deployer" as a managed service account name in each namespace
- [x] Stops injecting credentials as envvars to deployer pods
- [x] Uses mounted service account token as credential for deployer pod
- [x] Add check on startup to add default rolebindings to the default namespace
- [x] Make `osadm new-project` add bootstrap serviceaccount rolebindings by default
- [x] Coordinate tagging new deployer image to pick up service account token
- [x] Open issue for doc to remove `masterClients.deployerKubeConfig` https://github.com/openshift/openshift-docs/issues/451
- [x] Open issue for ansible to remove `masterClients.deployerKubeConfig` and references to `openshift-deployer.{crt,key,kubeconfig}` https://github.com/detiber/openshift-ansible/issues/37

liggitt commented:
[test]

liggitt commented:
@deads2k PTAL

liggitt commented:
@danmcp, fyi:
https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/2466/
```
Stderr from the command:

hack/test-cmd.sh: line 86: openshift: command not found
!!! Error in hack/test-cmd.sh:86
  'out=$(openshift version)' exited with status 127
Call stack:
  1: hack/test-cmd.sh:86 main(...)
Exiting with status 1
make: *** [test] Error 1
```

Re[test]

liggitt commented:
```
--- FAIL: TestConcurrentBuildControllers (23.25s)
	buildcontroller_test.go:103: The build was modified an unexpected number of times. Got: 0, Expected: 1
	buildcontroller_test.go:106: The build pod was created an unexpected number of times. Got: 0, Expected: 1
```
re[test]

liggitt commented:
re[test]

liggitt commented on hack/test-end-to-end.sh:
```diff
@@ -256,11 +256,23 @@ wait_for_url "${API_SCHEME}://${API_HOST}:${API_PORT}/api/v1beta1/minions/${KUBE
 openshift admin policy add-role-to-user view e2e-user --namespace=default
 
+# grant service accounts roles in the default namespace
+# TODO: figure out how to add these automatically to namespaces not created from project templates
```
This is gorpy, especially because the default namespace will not get created from a template, and we put stuff in there (registry, router, etc), so none of our service accounts will have permissions in that namespace without intervention like this. Any ideas?

deads2k replied:
> This is gorpy, especially because the default namespace will not get created from a template, and we put stuff in there (registry, router, etc), so none of our service accounts will have permissions in that namespace without intervention like this. Any ideas?

Don't use the default namespace?  This causes `osadm registry` to fail out of the box?

deads2k commented on Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/plugin/pkg/admission/serviceaccount/admission.go:
```diff
@@ -106,7 +106,7 @@ func NewServiceAccount(cl client.Interface) *serviceAccount {
 		Handler: admission.NewHandler(admission.Create),
 		// TODO: enable this once we've swept secret usage to account for adding secret references to service accounts
-		LimitSecretReferences: false,
+		LimitSecretReferences: true,
```
committed upstream?

deads2k commented:
I like this, but we can't merge it until we've sorted out how to grant permissions in a namespace.  We could make an "enable-namespace-for-openshift" command that uses a template to create our default role bindings.  `registry` and `router`, could then simply call the command as well and we'll work correctly again.

liggitt commented:
@deads2k added bindings to default namespace at startup (guarded by annotation), added bindings in `osadm new-project`, removed add-role-* workarounds from test-end-to-end.sh

deads2k commented on pkg/cmd/admin/project/new_project.go:
```diff
@@ -60,11 +62,12 @@ func NewCmdNewProject(name, fullName string, f *clientcmd.Factory, out io.Writer
+	cmd.Flags().StringVar(&options.DisplayName, "display-name", "", "Project display name")
+	cmd.Flags().StringVar(&options.Description, "description", "", "Project description")
 	cmd.Flags().StringVar(&options.NodeSelector, "node-selector", "", "Restrict pods onto nodes matching given label selector. Format: '<key1>=<value1>, <key2>=<value2>...'")
+	cmd.Flags().BoolVar(&options.AddDefaultRoles, "add-default-roles", true, "Add default bootstrap roles to project service accounts")
```
Why do we even have this lever?

deads2k commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -375,25 +376,51 @@ func GetBootstrapOpenshiftRoleBindings(openshiftNamespace string) []authorizatio
 
-func GetBootstrapClusterRoleBindings() []authorizationapi.ClusterRoleBinding {
-	return []authorizationapi.ClusterRoleBinding{
+func GetBootstrapServiceAccountRoleBindings(namespace string) []authorizationapi.RoleBinding {
```
This file is all about objects that are created once per cluster, not once per project.  The package is fine, but different file perhaps and something to indicate that these should be created for each project.

deads2k commented on pkg/cmd/server/api/validation/master.go:
```diff
@@ -184,6 +183,9 @@ func ValidateServiceAccountConfig(config api.ServiceAccountConfig, builtInKubern
 	if !managedNames.Has(bootstrappolicy.BuilderServiceAccountName) {
 		// TODO: warn that default builder service account won't be auto-created
 	}
+	if !managedNames.Has(bootstrappolicy.DeployerServiceAccountName) {
```
I added infrastructure for warnings, seems like this ought to actually produce a warning.

deads2k commented on pkg/project/registry/projectrequest/delegated/sample_template.go:
```diff
@@ -44,19 +43,11 @@ func DefaultTemplate() *templateapi.Template {
-	serviceAccountBuilderBinding.RoleRef.Name = bootstrappolicy.ImageBuilderRoleName
-	ret.Objects = append(ret.Objects, serviceAccountBuilderBinding)
+	for _, b := range bootstrappolicy.GetBootstrapServiceAccountRoleBindings(ns) {
+		// local copy to get unique pointer
```
reference by index?  That's done elsewhere in the code.

liggitt commented:
@deads2k comments addressed

deads2k commented:
lgtm [merge]

liggitt commented:
waiting to merge until I can coordinate with @smarterclayton on a tag to sync up the deployer docker image

liggitt commented:
[test]

liggitt commented:
Build push flake:
```
E0605 03:33:39.945627       1 dockerutil.go:44] Push for image 172.30.5.133:5000/default/origin-ruby-sample:latest failed, will retry in 10000000000s ...
E0605 03:34:09.953148       1 dockerutil.go:44] Push for image 172.30.5.133:5000/default/origin-ruby-sample:latest failed, will retry in 10000000000s ...
F0605 03:34:41.772040       1 builder.go:70] Build error: Failed to push image: API error (500):  v1 ping attempt failed with error: Get http://172.30.5.133:5000/v1/_ping: dial tcp 172.30.5.133:5000: i/o timeout
```

liggitt commented:
[test]

liggitt commented:
[test]

liggitt commented:
[merge]
